net: mqtt_sn: Improve handling of addresses

M1cha opened the pull request:
This PR fixes various issues I've found in the code related to addresses, especially when receiving multicast messages. There are more fixes coming up, but these would make this PR too big IMO.

All of these were found by reading the code, meaning I did not encounter these issues during my usage of the client. I still plan to write tests for the UDP transport, but I think discussing and fixing potential issues first and writing regression tests later is a good idea - especially since some commits change semantics of existing API.

M1cha committed:
net: mqtt_sn: fix types of gw_id and addr

- gw_id is an uint8_t in the protocol, so neither char, nor uint16_t are
  the correct type to use.
- addr is an array of bytes, not a string (of characters).

This also corrects related log messages and makes them more consistent, by
using hex for all gw_id's.

Signed-off-by: Michael Zimmermann <michael.zimmermann@sevenlab.de>

M1cha committed:
net: mqtt_sn: drop packets with truncated addresses

The address is only used for gateway discovery, but using a truncated
`struct sockaddr` can lead to out of bounds reads.

Signed-off-by: Michael Zimmermann <michael.zimmermann@sevenlab.de>

M1cha committed:
net: mqtt_sn: remove broadcast handling from recvfrom

I don't know, why this was added originally, but there are many issues with
it:
- You can't really return NULL, because src_addr is a `void*`, not a
  `void**`. The `src_addr = NULL` didn't do anything.
- The code was checking if the src_addr equals to the broadcast address,
  but under normal conditions, that's always a unicast address.
- Even if we would check if the destination address was the bcaddr, I don't
  see how that's useful. Currently, the address is only used for messages
  of types advertise and gwinfo, which use it to create a new entry in the
  gateway list. In that list, you always need the unicast address, never
  the broadcast address.

Signed-off-by: Michael Zimmermann <michael.zimmermann@sevenlab.de>

M1cha committed:
net: mqtt_sn: udp: use ssize_t for return value of zsock_recvfrom

int is simply the wrong type.

Signed-off-by: Michael Zimmermann <michael.zimmermann@sevenlab.de>

rlubos reviewed (dismissed):
Looks good, additional tests would be highly appreciated indeed.

M1cha commented on subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:
```diff
@@ -211,19 +211,19 @@ static ssize_t tp_udp_recvfrom(struct mqtt_sn_client *client, void *buffer, size
-	LOG_DBG("recv %d", rc);
-	if (rc < 0) {
+	ret = zsock_recvfrom(udp->sock, buffer, length, 0, src_addr, &addrlen_local);
+	LOG_DBG("recv %zd", ret);
```
I guess the idea was, that you log the number of bytes that have been read, since that's what `ret` contains if successful. Good point about errno though.

M1cha replied:
> Note that there is a problem with using ssize_t as a return value in this function as ssize_t is suppose to support only -1 and >=0 values. So returning -errno is "out of spec" here. 

wow, I didn't know that, thx. I'll think about the best way to fix this.

jukkar replied:
Handling a return from `zsock_recvfrom()` that would not fit to an int sounds very unrealistic in the zephyr, so I would not be very concerned about it.

M1cha replied:
IMO converting an ssize_t to an int based on that assumption is just as bad as assuming that ssize_t can store <-1. Personally, I prefer the latter since C doesn't even have a datatype which can only store a single negative value only and this seems to be common in Zephyr.

M1cha committed:
net: mqtt_sn: udp: backup errno if it's used in logs

Otherwise, the log might overwrite it.

Signed-off-by: Michael Zimmermann <michael.zimmermann@sevenlab.de>

M1cha commented:
I rebased.

M1cha committed:
net: mqtt_sn: udp: tp_udp_recvfrom: store error in errno variable

Returning negative values which are not -1, isn't what POSIX would do and
the API is specifically designed to behave like recvfrom.

Signed-off-by: Michael Zimmermann <michael.zimmermann@sevenlab.de>
